File: src/main/java/org/lanternpowered/server/event/LanternCauseStackManager.java
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.event;

import static org.lanternpowered.server.event.CauseStack.current;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import org.spongepowered.api.event.CauseStackManager;
import org.spongepowered.api.event.cause.Cause;
import org.spongepowered.api.event.cause.EventContext;
import org.spongepowered.api.event.cause.EventContextKey;

import java.util.Optional;

/**
 * A {@link CauseStackManager} that manages the {@link LanternCauseStack}s for all
 * the supported {@link Thread}s. (main, world threads, etc.)
 */
@Singleton
public class LanternCauseStackManager implements CauseStackManager {

    @Inject
    private LanternCauseStackManager() {
    }

    @Override
    public Cause getCurrentCause() {
        return current().getCurrentCause();
    }

    @Override
    public EventContext getCurrentContext() {
        return current().getCurrentContext();
    }

    @Override
    public CauseStackManager pushCause(Object obj) {
        return current().pushCause(obj);
    }

    @Override
    public Object popCause() {
        return current().popCause();
    }

    @Override
    public void popCauses(int n) {
        current().popCauses(n);
    }

    @Override
    public Object peekCause() {
        return current().peekCause();
    }

    @Override
    public StackFrame pushCauseFrame() {
        return current().pushCauseFrame();
    }

    @Override
    public void popCauseFrame(StackFrame handle) {
        current().popCauseFrame(handle);
    }

    @Override
    public <T> CauseStackManager addContext(EventContextKey<T> key, T value) {
        return current().addContext(key, value);
    }

    @Override
    public <T> Optional<T> getContext(EventContextKey<T> key) {
        return current().getContext(key);
    }

    @Override
    public <T> Optional<T> removeContext(EventContextKey<T> key) {
        return current().removeContext(key);
    }
}

File: src/main/kotlin/org/lanternpowered/server/cause/LanternCauseStackManager.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.cause

import org.lanternpowered.api.cause.Cause
import org.lanternpowered.api.cause.CauseContext
import org.lanternpowered.api.cause.CauseContextKey
import org.lanternpowered.api.cause.CauseStack
import org.lanternpowered.api.cause.CauseStackManager
import org.lanternpowered.api.util.concurrent.ThreadLocal
import org.lanternpowered.server.util.LanternThread

import java.util.Optional
import kotlin.reflect.KClass

/**
 * A [CauseStackManager] that manages the [LanternCauseStack]s for all
 * the supported [Thread]s. (main, world threads, etc.)
 */
object LanternCauseStackManager : CauseStackManager {

    /**
     * A [ThreadLocal] to fall back to if a thread doesn't extend [LanternThread].
     */
    private val fallbackCauseStacks = ThreadLocal<LanternCauseStack>()

    /**
     * Sets the [LanternCauseStack] for the current [Thread].
     */
    fun setCurrentCauseStack(causeStack: LanternCauseStack) {
        val thread = Thread.currentThread()
        if (thread is LanternThread) {
            thread.causeStack = causeStack
        } else {
            this.fallbackCauseStacks.set(causeStack)
        }
    }

    override fun currentStackOrNull() = this.getCauseStackOrNull(Thread.currentThread())

    private fun getCauseStackOrNull(thread: Thread)
            = if (thread is LanternThread) thread.causeStack else this.fallbackCauseStacks.get()

    fun getCauseStackOrEmpty(thread: Thread) = this.getCauseStackOrNull(thread) ?: EmptyCauseStack

    override fun currentStackOrEmpty(): CauseStack = this.currentStackOrNull() ?: EmptyCauseStack
    override fun currentStack() = this.currentStackOrNull() ?: error("The current thread doesn't support a cause stack.")

    override fun getCurrentCause(): Cause = this.currentStack().currentCause
    override fun getCurrentContext(): CauseContext = this.currentStack().currentContext
    override fun pushCause(obj: Any): CauseStackManager = apply { this.currentStack().pushCause(obj) }
    override fun popCause(): Any = this.currentStack().popCause()
    override fun popCauses(n: Int) = this.currentStack().popCauses(n)
    override fun peekCause(): Any = this.currentStack().peekCause()
    override fun pushCauseFrame(): CauseStack.Frame = this.currentStack().pushCauseFrame()
    override fun popCauseFrame(handle: org.spongepowered.api.event.CauseStackManager.StackFrame) = this.currentStack().popCauseFrame(handle)
    override fun <T : Any> addContext(key: CauseContextKey<T>, value: T): CauseStackManager = apply { this.currentStack().addContext(key, value) }
    override fun <T : Any> getContext(key: CauseContextKey<T>): Optional<T> = this.currentStack().getContext(key)
    override fun <T : Any> removeContext(key: CauseContextKey<T>): Optional<T> = this.currentStack().removeContext(key)
    override fun <T : Any> get(key: CauseContextKey<T>): T?= this.currentStack()[key]
    override fun <T : Any> set(key: CauseContextKey<T>, value: T) = this.currentStack().set(key, value)
    override fun <T : Any> first(target: KClass<T>): T?  = this.currentStack().first(target)
    override fun <T : Any> first(target: Class<T>): Optional<T> = this.currentStack().first(target)
    override fun <T : Any> last(target: KClass<T>): T? = this.currentStack().last(target)
    override fun <T : Any> last(target: Class<T>): Optional<T> = this.currentStack().last(target)
    override fun containsType(target: KClass<*>): Boolean = this.currentStack().containsType(target)
    override fun containsType(target: Class<*>): Boolean = this.currentStack().containsType(target)
    override fun contains(any: Any): Boolean = this.currentStack().contains(any)
}
